### PR TITLE
feat(zapslog): implement stack trace handling

### PR DESCRIPTION
**Pull request by zekth**
fix: https://github.com/uber-go/zap/issues/1329

As discussed in the issue, replacing the `HandlersOptions` with functional options, i also renamed `addSource` to `addCaller` to properly match the `zap` semantic.

### CODE REVIEW

File: stacktrace.go
Comment (zekth) [inline stacktrace.go:130]: Should we `deprecate` this and add a comment that it's just a temporary export?

File: exp/zapslog/slog_pre_go121.go
Comment (zekth) [inline exp/zapslog/slog_pre_go121.go:139]: `Stack` was not required for the implementation

File: stacktrace.go
Comment (abhinav) [reply stacktrace.go:130]: I don't like the idea of exporting this functionality from the main Zap package even as a temporary unsupported thing.
Instead, I've moved the relevant bits into an internal package in https://github.com/uber-go/zap/pull/1341.

That makes it easy to use them here.

**Review by abhinav — COMMENTED** [review]
Thanks for doing this, and thanks for waiting for the review. It's a bit busy these days.

I'll try to get https://github.com/uber-go/zap/pull/1341 landed soon to capture the stack traces without exporting them from the main Zap package.

And yeah, exp/ will need a `replace` directive. We'll have to be careful about updating exp/ to the latest Zap release when we tag it.

**Comment by abhinav** [comment]
I've pushed some minor fixups to your branch. Please be sure not to lose them if you `git push -f`.

File: stacktrace.go
Comment (zekth) [reply stacktrace.go:130]: Thanks for this i agree on this approach too

File: exp/zapslog/slog_go121.go
Comment (zekth) [reply exp/zapslog/slog_go121.go]: Sure I'll add revelant comments

**Comment by zekth** [comment]
@abhinav I rebased and applied all the comments

**Comment by zekth** [comment]
@abhinav should i add tests for the options or we add them in another PR? I'm happy to write a follow up one if needed

File: exp/zapslog/handler.go
Comment (abhinav) [inline exp/zapslog/handler.go:153]: Thanks for adding this. Much more obvious what's going on now.

File: exp/zapslog/options.go
Comment (abhinav) [inline exp/zapslog/options.go:59]: Minor outdated comment here. This is no Logger/SugaredLogger-based.

Plus this should probably be named WithCallerSkip to match WithCaller above.

File: exp/zapslog/options.go
Comment (abhinav) [inline exp/zapslog/options.go:67]: Just thinking out loud (would love your opinion on this, @zekth):
I wonder if this should be a slog Level instead. Or a new zapslog.LevelEnabler interface based on slog.Level.

```go
type LevelEnabler interface{ Enabled(context.Context, Level) bool }
```

The reasoning is this: the result of zapslog.NewHandler is a slog Handler. It talks in the language of slog.

**Review by abhinav — COMMENTED** [review]
Hey @zekth, apologies; I haven't forgotten about this, but I've had (and continue to have) a busy few days. I've taken a pass; this mostly looks good to go with minor comments. I'll address a few myself in a few minutes.

Tagging other Zap maintainers as well. CC @sywhang @mway @prashantv 
(Context: This PR turns HandlerOptions into functional options per prior discussion, and adds support for stack traces. Stack traces are captured for Error and above -- same as Zap's production logger.)

RE: tests: There's no need to backfill tests for existing options at this time--tests for the new option you're adding for stack traces should be sufficient. We do intend to fill in the rest of the tests before moving this to the main Zap package, though, but that's out of scope for this PR.

**Comment by abhinav** [comment]
On LevelEnabler and whether this should use a Zap level or slog level.
The more I think about it, the slog level makes sense.
The slog level has more information; conversion from slog to Zap level is a lossy operation.
I've renamed the option to AddStacktraceAt.
That way, if we want to add a LevelEnabler in the future, we can still use an `AddStacktrace(LevelEnabler)` option.
(The proposed LevelEnabler's Enabled signature matches slog.Handler.Enabled, so such a user-provided function
would have all the important information about the record to determine whether it should capture a stacktrace.)

**Review by abhinav — APPROVED** [review]
LGTM!

@mway @sywhang would appreciate a second look.
FYI, the coverage drop is because some of the existing functionality is untested; we will want to fix that before considering this stable anyway.

**Review by sywhang — APPROVED** [review]
LGTM.

**Pull request merged**
